src/engine/effects/engineeffectchain.cpp

```diff
@@ -128,113 +141,163 @@ bool EngineEffectChain::processEffectsRequest(const EffectsRequest& message,
     return true;
 }
 
-bool EngineEffectChain::enableForChannel(const ChannelHandle& handle) {
-    ChannelStatus& status = getChannelStatus(handle);
-    if (status.enable_state != EffectProcessor::ENABLED) {
-        status.enable_state = EffectProcessor::ENABLING;
+bool EngineEffectChain::enableForInputChannel(const ChannelHandle& inputHandle,
+        EffectStatesMapArray* statesForEffectsInChain) {
+    if (kEffectDebugOutput) {
+        qDebug() << "EngineEffectChain::enableForInputChannel" << this << inputHandle;
+    }
+    auto& outputMap = m_chainStatusForChannelMatrix[inputHandle];
+    for (auto&& outputChannelStatus : outputMap) {
+        VERIFY_OR_DEBUG_ASSERT(outputChannelStatus.enable_state !=
+                EffectEnableState::Enabled) {
+            for (auto&& pStatesMap : *statesForEffectsInChain) {
```

Comment:
Those 2 nested loops are unneeded. IMHO all destructors will be invoked when leaving the function's scope.

Reply:
See [earlier discussion](https://github.com/mixxxdj/mixxx/pull/1254#commitcomment-26321108) and latest commit

Reply:
Looks like I better go to bed now and review tomorrow :smile: Thanks for introducing std::unique_ptr for passing ownership, IMHO that's the safe and correct way 👍 If we really need to enclose a 4-element static array into a std::unqiue_ptr can be discussed. It shouldn't really matter to pass this around filled with null pointers. Two levels of unique_ptr might be confusing.